tests/v3/test_store/test_stateful_store.py

```diff
@@ -0,0 +1,232 @@
+# Stateful tests for arbitrary Zarr stores.
+
+
+import hypothesis.strategies as st
+from hypothesis import assume, note
+from hypothesis.stateful import (
+    RuleBasedStateMachine,
+    invariant,
+    precondition,
+    rule,
+)
+
+import zarr
+from zarr.abc.store import AccessMode, Store
+from zarr.core.buffer import Buffer, BufferPrototype, default_buffer_prototype
+from zarr.core.sync import sync
+from zarr.store import MemoryStore
+from zarr.testing.strategies import key_ranges, paths
+
+
+class SyncStoreWrapper(zarr.core.sync.SyncMixin):
+    def __init__(self, store: Store):
+        """Synchronous Store wrapper
+
+        This class holds synchronous methods that map to async methods of Store classes.
+        The synchronous wrapper is needed because hypothesis' stateful testing infra does
+        not support asyncio so we redefine sync versions of the Store API.
+        https://github.com/HypothesisWorks/hypothesis/issues/3712#issuecomment-1668999041
+        """
+        self.store = store
+
+    @property
+    def mode(self) -> AccessMode:
+        return self.store.mode
+
+    def set(self, key, data_buffer) -> None:
+        return sync(self.store.set(key, data_buffer))
+
+    def list(self) -> list:
+        async def wrapper(gen):
+            return [i async for i in gen]
+
+        gen = self.store.list()
+        yield from sync(wrapper(gen))
```

Comment:
I think this will work :)

```suggestion
    return self._sync_iter(self.store.list())
```